Windows - Add product, company name, and version info, for better windows file info display

NathanW2 opened the pull request:
## Description

Adds some version and product information in our exe and dll outputs for Windows so it's easier to check what version a library/exe is.

Also looks more pro and annoys me when stuff doesn't have it as it helps with debugging or checking which version a file came from. Good for IT auditing scripts as well.

Changes this PR makes:

- Add version and company info into the `dll` and `exe`
- Move the `.ico` files from inside the `app` folder that are only used on windows into `platform/windows`
- Add a version.rc template into  `platform/windows` for stamping the Windows build outputs with version info

## What does this look like?

![image](https://github.com/user-attachments/assets/49104c74-769e-4fdf-a026-7d644dac575b)

![image](https://github.com/user-attachments/assets/a40233d2-6da4-4126-91b1-0d723cbee904)

![image](https://github.com/user-attachments/assets/d2980fd3-5c1d-4953-98e5-1bd0b395c031)

and a more info in WinDbg

![image](https://github.com/user-attachments/assets/2e6ce1e0-bbd4-4677-91b9-49a2f2697ec9)

<img width="1199" height="674" alt="image" src="https://github.com/user-attachments/assets/bd684f63-70b6-4b41-80e2-7a24a4bf67d2" />

NathanW2 committed:
Windows - Add product, company name, and version into for display in Windows file Details tab and info

NathanW2 commented:
Might have a issue with static library type.  Checking on that. Didn't know we supported that.

NathanW2 committed:
Don't include rc file if static build type

agiudiceandrea commented:
See also: https://trac.osgeo.org/osgeo4w/ticket/888.

NathanW2 commented:
> See also: https://trac.osgeo.org/osgeo4w/ticket/888.

@agiudiceandrea good timing.

Yep that will work now

![image](https://github.com/user-attachments/assets/7b823e5c-4356-44fb-b12e-460df0cf7d7e)

baswein commented:
I have a noticed that my QGIS version in the start menu gets out of date when I update using OSGEO4W. Do you think this will help that? 
When I search for QGIS - I get records of versions that are not longer installed. 
![image](https://github.com/user-attachments/assets/8221d44b-060d-426d-9e73-f1d493671d0e)
This isn't a big deal because when I choose 3.32 it opens the latest version (3.42).  
![image](https://github.com/user-attachments/assets/d221012c-b0fa-4cb4-8a19-6a9fc41f5ff0)
Same for the LTR and the nightly.

NathanW2 commented:
No that is a different issue.  Will be related to the installer

On Sun, 6 July 2025, 12:28 am Sebastian Gutwein, ***@***.***>
wrote:

> *baswein* left a comment (qgis/QGIS#62516)
> <https://github.com/qgis/QGIS/pull/62516#issuecomment-3039009534>
>
> I have a noticed that my QGIS version in the start menu gets out of date
> when I update using OSGEO4W. Do you think this will help that?
> When I search for QGIS - I get records of versions that are not longer
> installed.
> image.png (view on web)
> <https://github.com/user-attachments/assets/8221d44b-060d-426d-9e73-f1d493671d0e>
> This isn't a big deal because when I choose 3.32 it opens the latest
> version (3.42).
> image.png (view on web)
> <https://github.com/user-attachments/assets/d221012c-b0fa-4cb4-8a19-6a9fc41f5ff0>
> Same for the LTR and the nightly.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/62516#issuecomment-3039009534>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC5FXFDWH3OLEFNUESO3333G7OKFAVCNFSM6AAAAACA26ZT36VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTAMZZGAYDSNJTGQ>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

agiudiceandrea commented:
@baswein, I guess it's more likely a MS Windows OS bug. Anyway it would be more effective to report such issue at https://trac.osgeo.org/osgeo4w/.

NathanW2 committed:
Split icon and version resource into different files

NathanW2 committed:
Move cmake functions into cmake file

NathanW2 commented:
Not sure why this builds locally but fails on Qt6 here..

m-kuhn commented on CMakeLists.txt:
```diff
@@ -75,6 +75,39 @@ message(STATUS "QGIS version: ${COMPLETE_VERSION} ${RELEASE_NAME} (${QGIS_VERSIO
+  set(${SRC_LIST} "${${SRC_LIST}}" PARENT_SCOPE)
+endfunction()
+
+# Function to create win32 VERSIONINFO and resource blocks
+# Only runs for WIN32 when called
+function(win32_version_info desc filename SRC_LIST)
```
I guess we can remove this (and the ones in CreateQgsVersion are kept)?

NathanW2 replied:
My main reason for keeping them out of the CREATE_QGSVERSION is because they are slightly different and I didn't want to risk breaking anything inside that macro

m-kuhn replied:
You also added a function with the same name in cmake/CreateQgsVersion.cmake (see https://github.com/qgis/QGIS/pull/62516/files#diff-ee874e32953e5696a8a87151cc69c723f92903edd0cc22a6cbfe20dc4978b744R78)

I would like to keep this out of the same CMakeLists.txt file, that file is huge already.

NathanW2 replied:
Oh yeah oops I was meant to remove that other one in the main cmake file
after I moved it. Sorry I misread the one you highlighted and messed up my commits.

m-kuhn commented:
> Not sure why this builds locally but fails on Qt6 here..

Are you also using ninja?

https://github.com/qgis/QGIS/blob/7b22b945954e65cd169e6e6920525e603aaa08c7/.github/workflows/windows-qt6.yml#L77

another difference might be ccache which is used

NathanW2 commented:
Oh yeah oops I was meant to remove that other one in the main cmake file
after I moved it. Sorry I misread the one you highlighted.

On Sun, 6 July 2025, 9:56 pm Matthias Kuhn, ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In CMakeLists.txt
> <https://github.com/qgis/QGIS/pull/62516#discussion_r2188214974>:
>
> > +# Function to create win32 VERSIONINFO and resource blocks
> +# Only runs for WIN32 when called
> +function(win32_version_info desc filename SRC_LIST)
>
> You also added a function with the same name in
> cmake/CreateQgsVersion.cmake (see
> https://github.com/qgis/QGIS/pull/62516/files#diff-ee874e32953e5696a8a87151cc69c723f92903edd0cc22a6cbfe20dc4978b744R78
> )
>
> I would like to keep this out of the main CMakeLists.txt file, that file
> is huge already.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/62516#discussion_r2188214974>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC5FXF5OU7VI64T5LSBOA33HEFIDAVCNFSM6AAAAACA26ZT36VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDSOJRGIYTAMJSGA>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

NathanW2 committed:
Remove extra cmake functions from main cmake

NathanW2 commented:
@m-kuhn yeah I'm using ninja but I haven't setup ccache yet on Windows.  I will see if I can find anything.  Some of the new resource files seem to work but not the gui one which is odd.

AScott-WWF commented:
Thanks for tagging my issue I raised on the OSGeo4W tracker - https://trac.osgeo.org/osgeo4w/ticket/888
I'd appreciate this being fixed if possible

NathanW2 commented:
Back from leave. So I will try and tackle this on the weekend so it is mergeable.  Might disable ccache on this PR for Qt6 to see if that is the issue.

NathanW2 committed:
Temp - Disable CCACHE to check if that is breaking this build

Disable CCACHE to check if that is breaking this build with new win32 resource files

NathanW2 committed:
Merge branch 'master' into win32_details

NathanW2 committed:
Put ccache back on

m-kuhn commented:
Is the file in ms-windows/rc/qgis.ico a copy of src/app/qgis.ico (and if yes, can we de-duplicate)?

m-kuhn commented on cmake/CreateQgsVersion.cmake:
```diff
@@ -58,3 +58,37 @@ MACRO(CREATE_QGSVERSION)
+
+  set(OUT "${PROJECT_SOURCE_DIR}/ms-windows/rc/icon.rc")
+  list(APPEND ${SRC_LIST} "${OUT}")
+  message(DEBUG "Win32 Icon Resource")
```
we can run cmake with `--trace` to discover this

NathanW2 replied:
@m-kuhn handy didn't know about that command.  Will fix it up.

m-kuhn committed:
Do not use ccache for rc files

m-kuhn committed:
only use ccache for c and cxx

NathanW2 commented:
> Is the file in ms-windows/rc/qgis.ico a copy of src/app/qgis.ico (and if yes, can we de-duplicate)?

Yeah it is.  I didn't remove the other one as I wasn't sure yet if other stuff needed it but I will check and remove if not needed.

m-kuhn committed:
use_ccache

m-kuhn committed:
Merge branch 'master' into win32_details

m-kuhn commented:
for reference: https://gitlab.kitware.com/cmake/cmake/-/issues/21328#note_846471

m-kuhn committed:
use CMAKE_[..]_COMPILER_LAUNCHER

NathanW2 committed:
Remove debug messages from cmake

NathanW2 committed:
Move windows resources into platform windows folder

NathanW2 committed:
Move windows ico files from app into platform/windows

NathanW2 committed:
Remove extra qgis ico file

NathanW2 commented:
@agiudiceandrea are you able to post a update on that  https://trac.osgeo.org/osgeo4w/ticket/888 I don't have a login at the moment.

ValentinBuira commented:
I have an error in building QGIS on linux(Ubuntu) since this commit. With an error related to ccache ? 

Here the tail of the error after running `ninja`, does that make sense to you ?

```
FAILED: src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader18.cpp.o 
"ccache sloppiness=pch_defines,time_macros" /usr/bin/c++ -DQT_DISABLE_DEPRECATED_BEFORE=0x050800 -DQT_NO_CAST_TO_ASCII -DQT_USE_QSTRINGBUILDER -DSIP_VERSION=0x060806 -DWITH_QTWEBKIT -I/home/valentin/Documents/QGIS/build-ninja/src/app/dwg/libdxfrw/libdxfrw_autogen/include -I/home/valentin/Documents/QGIS/build-ninja -I/home/valentin/Documents/QGIS/external/libdxfrw -isystem /home/valentin/Documents/QGIS/external/nlohmann -Wall -Wextra -Wno-long-long -Wformat-security -Wno-strict-aliasing -Wnon-virtual-dtor -Wno-redundant-move -Wno-misleading-indentation -Wno-deprecated-copy -g -fPIC -fvisibility=hidden -MD -MT src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader18.cpp.o -MF src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader18.cpp.o.d -o src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader18.cpp.o -c /home/valentin/Documents/QGIS/external/libdxfrw/intern/dwgreader18.cpp
/bin/sh: 1: ccache sloppiness=pch_defines,time_macros: not found
[18/4460] Building CXX object src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader21.cpp.o
FAILED: src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader21.cpp.o 
"ccache sloppiness=pch_defines,time_macros" /usr/bin/c++ -DQT_DISABLE_DEPRECATED_BEFORE=0x050800 -DQT_NO_CAST_TO_ASCII -DQT_USE_QSTRINGBUILDER -DSIP_VERSION=0x060806 -DWITH_QTWEBKIT -I/home/valentin/Documents/QGIS/build-ninja/src/app/dwg/libdxfrw/libdxfrw_autogen/include -I/home/valentin/Documents/QGIS/build-ninja -I/home/valentin/Documents/QGIS/external/libdxfrw -isystem /home/valentin/Documents/QGIS/external/nlohmann -Wall -Wextra -Wno-long-long -Wformat-security -Wno-strict-aliasing -Wnon-virtual-dtor -Wno-redundant-move -Wno-misleading-indentation -Wno-deprecated-copy -g -fPIC -fvisibility=hidden -MD -MT src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader21.cpp.o -MF src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader21.cpp.o.d -o src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader21.cpp.o -c /home/valentin/Documents/QGIS/external/libdxfrw/intern/dwgreader21.cpp
/bin/sh: 1: ccache sloppiness=pch_defines,time_macros: not found
[24/4460] Generating qgsversion.h, qgsversion.inc
ninja: build stopped: subcommand failed.

```

elpaso commented:
> I have an error in building QGIS on linux(Ubuntu) since this commit. With an error related to ccache ?
> 
> Here the tail of the error after running `ninja`, does that make sense to you ?
> 
> ```
> FAILED: src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader18.cpp.o 
> "ccache sloppiness=pch_defines,time_macros" /usr/bin/c++ -DQT_DISABLE_DEPRECATED_BEFORE=0x050800 -DQT_NO_CAST_TO_ASCII -DQT_USE_QSTRINGBUILDER -DSIP_VERSION=0x060806 -DWITH_QTWEBKIT -I/home/valentin/Documents/QGIS/build-ninja/src/app/dwg/libdxfrw/libdxfrw_autogen/include -I/home/valentin/Documents/QGIS/build-ninja -I/home/valentin/Documents/QGIS/external/libdxfrw -isystem /home/valentin/Documents/QGIS/external/nlohmann -Wall -Wextra -Wno-long-long -Wformat-security -Wno-strict-aliasing -Wnon-virtual-dtor -Wno-redundant-move -Wno-misleading-indentation -Wno-deprecated-copy -g -fPIC -fvisibility=hidden -MD -MT src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader18.cpp.o -MF src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader18.cpp.o.d -o src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader18.cpp.o -c /home/valentin/Documents/QGIS/external/libdxfrw/intern/dwgreader18.cpp
> /bin/sh: 1: ccache sloppiness=pch_defines,time_macros: not found
> [18/4460] Building CXX object src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader21.cpp.o
> FAILED: src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader21.cpp.o 
> "ccache sloppiness=pch_defines,time_macros" /usr/bin/c++ -DQT_DISABLE_DEPRECATED_BEFORE=0x050800 -DQT_NO_CAST_TO_ASCII -DQT_USE_QSTRINGBUILDER -DSIP_VERSION=0x060806 -DWITH_QTWEBKIT -I/home/valentin/Documents/QGIS/build-ninja/src/app/dwg/libdxfrw/libdxfrw_autogen/include -I/home/valentin/Documents/QGIS/build-ninja -I/home/valentin/Documents/QGIS/external/libdxfrw -isystem /home/valentin/Documents/QGIS/external/nlohmann -Wall -Wextra -Wno-long-long -Wformat-security -Wno-strict-aliasing -Wnon-virtual-dtor -Wno-redundant-move -Wno-misleading-indentation -Wno-deprecated-copy -g -fPIC -fvisibility=hidden -MD -MT src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader21.cpp.o -MF src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader21.cpp.o.d -o src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader21.cpp.o -c /home/valentin/Documents/QGIS/external/libdxfrw/intern/dwgreader21.cpp
> /bin/sh: 1: ccache sloppiness=pch_defines,time_macros: not found
> [24/4460] Generating qgsversion.h, qgsversion.inc
> ninja: build stopped: subcommand failed.
> ```

Same here, this is the source of the issue
https://github.com/qgis/QGIS/pull/62516/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR158

the error comes from the double quotes around the ccache invokation, if they are removed the command works.

NathanW2 commented:
Will remove and open a new PR to see if passes on all platforms.

On Wed, 6 Aug 2025, 6:46 pm Nathan Woodrow, ***@***.***> wrote:

> Hmmmm odd.
>
> On Wed, 6 Aug 2025, 6:22 pm Alessandro Pasotti, ***@***.***>
> wrote:
>
>> *elpaso* left a comment (qgis/QGIS#62516)
>> <https://github.com/qgis/QGIS/pull/62516#issuecomment-3158142614>
>>
>> I have an error in building QGIS on linux(Ubuntu) since this commit. With
>> an error related to ccache ?
>>
>> Here the tail of the error after running ninja, does that make sense to
>> you ?
>>
>> FAILED: src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader18.cpp.o
>> "ccache sloppiness=pch_defines,time_macros" /usr/bin/c++ -DQT_DISABLE_DEPRECATED_BEFORE=0x050800 -DQT_NO_CAST_TO_ASCII -DQT_USE_QSTRINGBUILDER -DSIP_VERSION=0x060806 -DWITH_QTWEBKIT -I/home/valentin/Documents/QGIS/build-ninja/src/app/dwg/libdxfrw/libdxfrw_autogen/include -I/home/valentin/Documents/QGIS/build-ninja -I/home/valentin/Documents/QGIS/external/libdxfrw -isystem /home/valentin/Documents/QGIS/external/nlohmann -Wall -Wextra -Wno-long-long -Wformat-security -Wno-strict-aliasing -Wnon-virtual-dtor -Wno-redundant-move -Wno-misleading-indentation -Wno-deprecated-copy -g -fPIC -fvisibility=hidden -MD -MT src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader18.cpp.o -MF src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader18.cpp.o.d -o src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader18.cpp.o -c /home/valentin/Documents/QGIS/external/libdxfrw/intern/dwgreader18.cpp
>> /bin/sh: 1: ccache sloppiness=pch_defines,time_macros: not found
>> [18/4460] Building CXX object src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader21.cpp.o
>> FAILED: src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader21.cpp.o
>> "ccache sloppiness=pch_defines,time_macros" /usr/bin/c++ -DQT_DISABLE_DEPRECATED_BEFORE=0x050800 -DQT_NO_CAST_TO_ASCII -DQT_USE_QSTRINGBUILDER -DSIP_VERSION=0x060806 -DWITH_QTWEBKIT -I/home/valentin/Documents/QGIS/build-ninja/src/app/dwg/libdxfrw/libdxfrw_autogen/include -I/home/valentin/Documents/QGIS/build-ninja -I/home/valentin/Documents/QGIS/external/libdxfrw -isystem /home/valentin/Documents/QGIS/external/nlohmann -Wall -Wextra -Wno-long-long -Wformat-security -Wno-strict-aliasing -Wnon-virtual-dtor -Wno-redundant-move -Wno-misleading-indentation -Wno-deprecated-copy -g -fPIC -fvisibility=hidden -MD -MT src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader21.cpp.o -MF src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader21.cpp.o.d -o src/app/dwg/libdxfrw/CMakeFiles/libdxfrw.dir/intern/dwgreader21.cpp.o -c /home/valentin/Documents/QGIS/external/libdxfrw/intern/dwgreader21.cpp
>> /bin/sh: 1: ccache sloppiness=pch_defines,time_macros: not found
>> [24/4460] Generating qgsversion.h, qgsversion.inc
>> ninja: build stopped: subcommand failed.
>>
>> Same here, this is the source of the issue
>>
>> https://github.com/qgis/QGIS/pull/62516/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR158
>>
>> the error comes from the double quotes around the ccache invokation, if
>> they are removed the command works.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/qgis/QGIS/pull/62516#issuecomment-3158142614>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAC5FXBMOEOT62FWU2TVJI33MG3LHAVCNFSM6AAAAACA26ZT36VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTCNJYGE2DENRRGQ>
>> .
>> You are receiving this because you modified the open/close state.Message
>> ID: ***@***.***>
>>
>
